File: inventario/src/DAO/InventarioDAO.java
package DAO;

import CX.ConexionBD;
import Modelo.MovimientoInventario;
import Utils.AESUtil;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class InventarioDAO {

    public List<MovimientoInventario> listarMovimientos() throws SQLException {
        List<MovimientoInventario> lista = new ArrayList<>();
        String sql = "SELECT i.* FROM inventario i " +
                     "ORDER BY i.id DESC";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) lista.add(mapear(rs));
        }
        return lista;
    }

    public List<MovimientoInventario> listarMovimientosConNombres() throws SQLException {
        List<MovimientoInventario> lista = new ArrayList<>();
        String sql = "SELECT i.*, p.nombre as producto_nombre, pr.nombre_empresa as proveedor_nombre " +
                     "FROM inventario i " +
                     "LEFT JOIN producto p ON i.producto_id = p.id " +
                     "LEFT JOIN proveedor pr ON i.proveedor_id = pr.id " +
                     "ORDER BY i.id DESC";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql);
             ResultSet rs = ps.executeQuery()) {
            while (rs.next()) lista.add(mapear(rs));
        }
        return lista;
    }

    public List<MovimientoInventario> buscarPorProducto(String texto) throws SQLException {
        // Simplificado para evitar JOINs complejos con nombres encriptados que requieren lógica de búsqueda distinta
        List<MovimientoInventario> lista = new ArrayList<>();
        for (MovimientoInventario m : listarMovimientos()) {
            if (m.getProductoNombre() != null && m.getProductoNombre().toLowerCase().contains(texto.toLowerCase())) {
                lista.add(m);
            }
        }
        return lista;
    }

    public long registrarMovimiento(MovimientoInventario m) throws SQLException {
        String sqlMov = "INSERT INTO inventario (producto_id, proveedor_id, precio, precio_balance, cantidad, tipo_movimiento, motivo) " +
                        "VALUES (?, ?, ?, ?, ?, ?, ?)";
        String tipo = m.getTipoMovimiento();
        String sqlProd;
        double ajuste;
        if ("Ajuste".equals(tipo)) {
            sqlProd = "UPDATE producto SET stock_actual = ? WHERE id = ?";
            ajuste = m.getCantidad();
        } else {
            sqlProd = "UPDATE producto SET stock_actual = stock_actual + ? WHERE id = ?";
            ajuste = "Entrada".equals(tipo) ? m.getCantidad() : -m.getCantidad();
        }

        try (Connection con = ConexionBD.conectar()) {
            con.setAutoCommit(false);
            try {
                long id = -1;
                try (PreparedStatement ps = con.prepareStatement(sqlMov, Statement.RETURN_GENERATED_KEYS)) {
                    ps.setLong(1, m.getProductoId());
                    if (m.getProveedorId() != null) ps.setLong(2, m.getProveedorId());
                    else ps.setNull(2, Types.BIGINT);
                    ps.setDouble(3, m.getPrecio());
                    ps.setDouble(4, m.getPrecioBalance());
                    ps.setDouble(5, m.getCantidad());
                    ps.setString(6, AESUtil.encriptar(tipo));
                    ps.setString(7, AESUtil.encriptar(m.getMotivo()));
                    ps.executeUpdate();
                    try (ResultSet keys = ps.getGeneratedKeys()) {
                        if (keys.next()) id = keys.getLong(1);
                    }
                }

                try (PreparedStatement ps = con.prepareStatement(sqlProd)) {
                    ps.setDouble(1, ajuste);
                    ps.setLong(2, m.getProductoId());
                    ps.executeUpdate();
                }

                con.commit();
                return id;
            } catch (SQLException e) {
                con.rollback();
                throw e;
            }
        }
    }

    public double obtenerStockProducto(long productoId) throws SQLException {
        String sql = "SELECT stock_actual FROM producto WHERE id=?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setLong(1, productoId);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) return rs.getDouble("stock_actual");
            }
        }
        return 0;
    }

    public double obtenerVentasMes() throws SQLException {
        String encAnulada = AESUtil.encriptar("Anulada");
        String sql = "SELECT COALESCE(SUM(f.total), 0) FROM factura f " +
                     "WHERE MONTH(f.fecha_emision)=MONTH(CURRENT_DATE()) " +
                     "AND YEAR(f.fecha_emision)=YEAR(CURRENT_DATE()) " +
                     "AND f.estado != ?";
        try (Connection con = ConexionBD.conectar();
             PreparedStatement ps = con.prepareStatement(sql)) {
            ps.setString(1, encAnulada);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) return rs.getDouble(1);
            }
        }
        return 0;
    }

    private MovimientoInventario mapear(ResultSet rs) throws SQLException {
        MovimientoInventario m = new MovimientoInventario();
        m.setId(rs.getLong("id"));
        m.setProductoId(rs.getLong("producto_id"));
        long provId = rs.getLong("proveedor_id");
        m.setProveedorId(rs.wasNull() ? null : provId);
        m.setPrecio(rs.getDouble("precio"));
        m.setPrecioBalance(rs.getDouble("precio_balance"));
        m.setCantidad(rs.getDouble("cantidad"));
        m.setTipoMovimiento(AESUtil.desencriptar(rs.getString("tipo_movimiento")));
        m.setFechaMovimiento(rs.getTimestamp("fecha_movimiento"));
        m.setMotivo(AESUtil.desencriptar(rs.getString("motivo")));
        // Intentar obtener nombres si existen en el ResultSet (JOINs)
        try {
            m.setProductoNombre(AESUtil.desencriptar(rs.getString("producto_nombre")));
        } catch (Exception e) {}
        try {
            m.setProveedorNombre(AESUtil.desencriptar(rs.getString("proveedor_nombre")));
        } catch (Exception e) {}
        return m;
    }
}

File: inventario_python/dao/inventario_dao.py
from db.conexion import conectar
from models.movimiento_inventario import MovimientoInventario
from utils.aes_util import encriptar, desencriptar


class InventarioDAO:

    def _mapear(self, row):
        m = MovimientoInventario()
        m.id = row[0]
        m.producto_id = row[1]
        m.proveedor_id = row[2]
        m.precio = row[3] or 0.0
        m.precio_balance = row[4] or 0.0
        m.cantidad = row[5] or 0.0
        m.tipo_movimiento = desencriptar(row[6]) if row[6] else ""
        m.fecha_movimiento = row[7]
        m.motivo = desencriptar(row[8]) if row[8] else ""
        if len(row) > 9 and row[9]:
            m.producto_nombre = desencriptar(row[9])
        if len(row) > 10 and row[10]:
            m.proveedor_nombre = desencriptar(row[10])
        return m

    def listar_movimientos(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute("SELECT * FROM inventario ORDER BY id DESC")
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def listar_movimientos_con_nombres(self):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            cur.execute(
                "SELECT i.*, p.nombre as producto_nombre, pr.nombre_empresa as proveedor_nombre "
                "FROM inventario i "
                "LEFT JOIN producto p ON i.producto_id = p.id "
                "LEFT JOIN proveedor pr ON i.proveedor_id = pr.id "
                "ORDER BY i.id DESC"
            )
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def buscar_por_producto(self, texto):
        conn = conectar()
        if not conn:
            return []
        try:
            cur = conn.cursor()
            enc = encriptar(texto)
            cur.execute(
                "SELECT i.* FROM inventario i JOIN producto p ON i.producto_id = p.id "
                "WHERE p.nombre LIKE %s OR p.nombre = %s ORDER BY i.id DESC",
                (f"%{texto}%", enc),
            )
            return [self._mapear(r) for r in cur.fetchall()]
        finally:
            conn.close()

    def registrar_movimiento(self, m):
        conn = conectar()
        if not conn:
            return False
        try:
            cur = conn.cursor()
            tipo_enc = encriptar(m.tipo_movimiento) if m.tipo_movimiento else None
            motivo_enc = encriptar(m.motivo) if m.motivo else None
            cur.execute(
                "INSERT INTO inventario (producto_id, proveedor_id, precio, precio_balance, cantidad, tipo_movimiento, motivo) "
                "VALUES (%s,%s,%s,%s,%s,%s,%s)",
                (m.producto_id, m.proveedor_id, m.precio, m.precio_balance, m.cantidad, tipo_enc, motivo_enc),
            )
            if m.tipo_movimiento == "Entrada":
                cur.execute("UPDATE producto SET stock_actual = stock_actual + %s WHERE id = %s", (m.cantidad, m.producto_id))
            elif m.tipo_movimiento == "Salida":
                cur.execute("UPDATE producto SET stock_actual = stock_actual - %s WHERE id = %s", (m.cantidad, m.producto_id))
            elif m.tipo_movimiento == "Ajuste":
                cur.execute("UPDATE producto SET stock_actual = %s WHERE id = %s", (m.cantidad, m.producto_id))
            conn.commit()
            return True
        except Exception as e:
            conn.rollback()
            raise e
        finally:
            conn.close()

    def stock_actual(self, producto_id):
        conn = conectar()
        if not conn:
            return 0.0
        try:
            cur = conn.cursor()
            cur.execute("SELECT stock_actual FROM producto WHERE id = %s", (producto_id,))
            row = cur.fetchone()
            return row[0] if row else 0.0
        finally:
            conn.close()

    def ventas_del_mes(self):
        conn = conectar()
        if not conn:
            return 0.0
        try:
            cur = conn.cursor()
            anulada = encriptar("Anulada")
            cur.execute("SELECT COALESCE(SUM(total),0) FROM factura WHERE MONTH(fecha_emision) = MONTH(CURDATE()) AND YEAR(fecha_emision) = YEAR(CURDATE()) AND estado != %s", (anulada,))
            return cur.fetchone()[0]
        finally:
            conn.close()
